src/query/src/range_select/plan_rewrite.rs

```diff
@@ -0,0 +1,466 @@
+                        !range_plan.schema().has_column(column)
+                    } else {
+                        true
+                    }
+                }) || new_expr.len() != range_plan.schema().fields().len();
+                if need_project {
+                    let project_plan = LogicalPlanBuilder::from(range_plan)
+                        .project(new_expr)
+                        .context(DataFusionSnafu)?
+                        .build()
+                        .context(DataFusionSnafu)?;
+                    Ok(Some(project_plan))
+                } else {
+                    Ok(Some(range_plan))
+                }
+            }
+            _ => {
+                if new_inputs.iter().any(|x| x.is_some()) {
+                    let inputs: Vec<LogicalPlan> = new_inputs
+                        .into_iter()
+                        .zip(inputs)
+                        .map(|(x, y)| match x {
+                            Some(plan) => plan,
+                            None => y.clone(),
+                        })
+                        .collect();
+                    Ok(Some(
+                        plan.with_new_inputs(&inputs).context(DataFusionSnafu)?,
+                    ))
+                } else {
+                    Ok(None)
+                }
+            }
+        }
+    }
+
+    /// this function use to find the time_index column and row columns from input schema,
+    /// return `(time_index, [row_columns])` to the rewriter.
+    /// If the user does not explicitly use the `by` keyword to indicate time series,
+    /// `[row_columns]` will be use as default time series
+    async fn get_index_by(&mut self, schema: Arc<DFSchema>) -> Result<(Expr, Vec<Expr>)> {
+        let mut time_index_expr = Expr::Wildcard;
+        let mut default_by = vec![];
+        for field in schema.fields() {
+            if let Some(table_ref) = field.qualifier() {
+                let table = self
+                    .table_provider
+                    .resolve_table(table_ref.clone())
+                    .await
+                    .context(CatalogSnafu)?
+                    .as_any()
+                    .downcast_ref::<DefaultTableSource>()
+                    .context(UnknownTableSnafu)?
+                    .table_provider
+                    .as_any()
+                    .downcast_ref::<DfTableProviderAdapter>()
+                    .context(UnknownTableSnafu)?
+                    .table();
+                let schema = table.schema();
+                let time_index_column =
+                    schema
+                        .timestamp_column()
+                        .with_context(|| TimeIndexNotFoundSnafu {
+                            table: table_ref.to_string(),
+                        })?;
+                // assert time_index's datatype is timestamp
+                if let ConcreteDataType::Timestamp(datatypes::types::TimestampType::Millisecond(
```

Comment:
question: Do we need to support other resolutions? e.g. seconds

Reply:
Here my implementation refers to the greptimedb PromQL query part, because in the implementation of PromQL, the time index column will be forcibly cast to `TimestampMillisecondArray`. If we do not check whether the data is `TimestampMillisecond` here, there will be problems in the subsequent execution of the physical plan.